File: src/politi/tergm.py
"""Panel of two-mode networks for a temporal ERGM, and its design constraints.

A TERGM is estimable on this dataset, but three properties of it decide the
specification, and every one of them is a reason to narrow the claim rather
than widen it.

**The model must be fitted to the two-mode network.** Not to a projection. A
one-mode projection of an affiliation network manufactures a clique out of
every board, so its triangles and its clustering are arithmetic rather than
behaviour, and the dependence an ERGM is meant to estimate is already baked in
by the construction. Interlocking is represented here by `b2star(2)` — two
directors sharing a firm — on the two-mode graph, which is the same quantity
without the manufactured dependence.

**Node turnover is severe, and the first transition is thin.** A TERGM's
memory term lives on the dyads whose *both* endpoints exist in consecutive
waves. That set is 44 persons × 109 firms across 1932→1938 and 459 × 427
across 1947→1950 — a fortyfold difference. 1932 is also a selection of
prominent administrators rather than a full roster, so the first transition
mixes a change in the annuaire with a change in the network. Every model is
reported with and without it.

**The intervals are unequal — 6, 4, 5 and 3 years.** A TERGM has no offset for
that, so the memory coefficient is an average over four different gaps and is
not a per-year persistence rate. `timecov` lets it vary by transition, which
is the most that can be done here.

One discipline runs through the covariates. **A covariate computed from the
network cannot go on the right-hand side.** Seat counts, financier status and
centrality are all functions of the ties being modelled; including them would
regress the network on itself. Only attributes read off the page are used:
origin (imputed from the name), rank and public office on the person side,
sector on the firm side. Office is printed in the same entry as the
directorships, so it is contemporaneous and stays associational.
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd

#: Roles that count as a board seat, as in `network.BOARD_ROLES`.
from .network import BOARD_ROLES

#: Terms in the specification, and what each one is for.
SPECIFICATION = """\
edges                      baseline density
memory(autoregression)     a seat held at t-1, held again at t
b1star(2)                  a director holding two seats
b2star(2)                  two directors sharing a firm — interlocking
b1factor(office)           office holders' propensity to hold seats
b1factor(rank)             Pasha and Bey
b1nodematch(origin)        directors sharing a board with their own community
b2factor(sector)           finance, land, agriculture against the rest
"""


def network_panel(processed: Path | None = None,
                  drop_1932: bool = False) -> dict:
    """Edges, node sets and node attributes for each wave.

    Returns a dict with `waves`, `edges` (one frame per wave), `persons` and
    `firms` (attribute frames per wave). Only board roles are kept, and only
    records that pass `origin.is_person`.
    """
    from . import config
    from .origin import is_person
    from .sectors import sector

    processed = Path(processed) if processed else config.PROCESSED
    aff = pd.read_csv(processed / "affiliations.csv")
    aff = aff[aff.person_label.map(is_person) & aff.role.isin(BOARD_ROLES)]

    offices = processed / "person_political.csv"
    flags = (pd.read_csv(offices)[["year", "person_id", "political"]]
             if offices.exists() else
             pd.DataFrame(columns=["year", "person_id", "political"]))

    coded = _origin_by_person(aff)
    waves = sorted(aff.year.unique())
    if drop_1932:
        waves = [w for w in waves if w != 1932]

    edges, persons, firms = {}, {}, {}
    for year in waves:
        chunk = aff[aff.year == year]
        edges[year] = (chunk[["person_id", "company_id"]]
                       .drop_duplicates().reset_index(drop=True))
        p = (chunk[["person_id", "person_label", "rank"]].drop_duplicates("person_id")
             .merge(coded, on="person_id", how="left")
             .merge(flags[flags.year == year][["person_id", "political"]],
                    on="person_id", how="left"))
        p["political"] = p.political.fillna(False).astype(bool)
        p["origin"] = p.origin.fillna("unknown")
        p["rank"] = p["rank"].fillna("").replace("", "untitled")
        persons[year] = p.sort_values("person_id").reset_index(drop=True)

        f = chunk[["company_id", "company_label"]].drop_duplicates("company_id").copy()
        f["sector"] = f.company_label.map(sector)
        firms[year] = f.sort_values("company_id").reset_index(drop=True)
    return {"waves": waves, "edges": edges, "persons": persons, "firms": firms}


def _origin_by_person(aff: pd.DataFrame) -> pd.DataFrame:
    """One origin per person, from the canonical label.

    Origin is a property of the man, not of the wave, so it is coded once and
    held fixed — a name-based imputation that changed between volumes would be
    a linkage artefact, not a biography.
    """
    from .origin import classify_frame

    labels = aff.drop_duplicates("person_id")[["person_id", "person_label"]]
    coded = classify_frame(labels.person_label.unique())
    return (labels.merge(coded, left_on="person_label", right_on="label",
                         how="left")[["person_id", "origin"]])


def transition_table(panel: dict) -> pd.DataFrame:
    """What each transition has to work with.

    The memory term is estimated on dyads whose both endpoints appear in
    consecutive waves, so this table is the first thing to read before any
    coefficient: it is the model's actual sample size.
    """
    waves = panel["waves"]
    rows = []
    for a, b in zip(waves, waves[1:]):
        pa = set(panel["persons"][a].person_id)
        pb = set(panel["persons"][b].person_id)
        fa = set(panel["firms"][a].company_id)
        fb = set(panel["firms"][b].company_id)
        both_p, both_f = pa & pb, fa & fb
        ea = set(map(tuple, panel["edges"][a].to_numpy()))
        eb = set(map(tuple, panel["edges"][b].to_numpy()))
        at_risk = {(p, f) for p in both_p for f in both_f}
        ea_r = ea & at_risk
        eb_r = eb & at_risk
        rows.append({
            "from": a, "to": b, "gap_years": b - a,
            "persons_both": len(both_p), "firms_both": len(both_f),
            "dyads_at_risk": len(at_risk),
            "edges_from": len(ea_r), "edges_to": len(eb_r),
            "stable": len(ea_r & eb_r),
            "formed": len(eb_r - ea_r), "dissolved": len(ea_r - eb_r),
        })
    return pd.DataFrame(rows)


def export_for_r(panel: dict, outdir: Path) -> list[Path]:
    """Write the panel as CSVs for `scripts/tergm.R`.

    Plain CSV rather than an .RData blob, so the input to the model is
    readable and diffable in the repository.
    """
    outdir = Path(outdir)
    outdir.mkdir(parents=True, exist_ok=True)
    written = []
    for kind in ("edges", "persons", "firms"):
        frames = []
        for year in panel["waves"]:
            frame = panel[kind][year].copy()
            frame.insert(0, "year", year)
            frames.append(frame)
        path = outdir / f"tergm_{kind}.csv"
        pd.concat(frames, ignore_index=True).to_csv(path, index=False)
        written.append(path)
    return written


# --- estimation ---------------------------------------------------------------
#
# `btergm`'s default estimator is maximum pseudolikelihood with bootstrapped
# confidence intervals (Desmarais and Cranmer 2012; Leifeld, Cranmer and
# Desmarais 2018). That is a logistic regression on the at-risk dyads with
# each term's change statistic as a regressor, and a bootstrap that resamples
# *nodes* rather than dyads. Both are implemented here so the model can be
# fitted without R, and so the R result has something to be checked against.
#
# What this is not: MCMC-MLE. For that, use `scripts/tergm.R`.

#: Change statistics computed by :func:`change_statistics`, in order.
TERMS = ["edges", "memory", "b1star2", "b2star2", "office", "origin_match",
         "sector_finance", "sector_land_property", "sector_agriculture"]

#: Also computed, but not in the default specification. Pass it to
#: :func:`fit_mple` to ask whether an office holder's seat persists better
#: than anyone else's — which is a different question from whether he holds
#: more seats, and has a different answer. See `docs/TERGM.md`.
INTERACTIONS = ["memory_x_office"]


def change_statistics(panel: dict) -> pd.DataFrame:
    """One row per at-risk dyad per transition, with each term's change stat.

    The change statistic of a term is how much the network statistic moves
    when the dyad is toggled on, holding the rest of the observed network
    fixed — which is exactly what conditioning on the rest of the network
    means, and what makes this a pseudolikelihood rather than a likelihood.

    Only dyads whose both endpoints appear in the wave *and* the one before it
    are included, so `memory` distinguishes "no tie" from "did not exist".
    """
    waves = panel["waves"]
    rows = []
    for previous, current in zip(waves, waves[1:]):
        persons = panel["persons"][current]
        firms = panel["firms"][current]
        prev_p = set(panel["persons"][previous].person_id)
        prev_f = set(panel["firms"][previous].company_id)

        at_risk_p = persons[persons.person_id.isin(prev_p)].reset_index(drop=True)
        at_risk_f = firms[firms.company_id.isin(prev_f)].reset_index(drop=True)
        if at_risk_p.empty or at_risk_f.empty:
            continue

        now = set(map(tuple, panel["edges"][current].to_numpy()))
        before = set(map(tuple, panel["edges"][previous].to_numpy()))

        origin = dict(zip(at_risk_p.person_id, at_risk_p.origin))
        office = dict(zip(at_risk_p.person_id, at_risk_p.political))
        sectors = dict(zip(at_risk_f.company_id, at_risk_f.sector))

        # Degrees and the origin composition of each board, on the observed
        # network at t, restricted to the at-risk node set.
        p_ids = list(at_risk_p.person_id)
        f_ids = list(at_risk_f.company_id)
        p_set, f_set = set(p_ids), set(f_ids)
        p_degree = dict.fromkeys(p_ids, 0)
        f_degree = dict.fromkeys(f_ids, 0)
        board_origins: dict[str, dict[str, int]] = {f: {} for f in f_ids}
        for person, firm in now:
            if person in p_set and firm in f_set:
                p_degree[person] += 1
                f_degree[firm] += 1
                who = origin.get(person, "unknown")
                if who != "unknown":
                    board_origins[firm][who] = board_origins[firm].get(who, 0) + 1

        for person in p_ids:
            ego = origin.get(person, "unknown")
            has_office = bool(office.get(person, False))
            for firm in f_ids:
                tie = (person, firm) in now
                # Degrees excluding the focal tie: the change statistic of a
                # two-star is the partner count the new tie would join.
                deg_p = p_degree[person] - (1 if tie else 0)
                deg_f = f_degree[firm] - (1 if tie else 0)
                same = 0
                if ego != "unknown":
                    same = board_origins[firm].get(ego, 0) - (1 if tie else 0)
                rows.append((
                    current, person, firm, int(tie),
                    1,                                        # edges
                    int((person, firm) in before),            # memory
                    deg_p, deg_f,
                    int(has_office),
                    same,
                    int(sectors.get(firm) == "finance"),
                    int(sectors.get(firm) == "land_property"),
                    int(sectors.get(firm) == "agriculture"),
                    int((person, firm) in before) * int(has_office),
                ))
    return pd.DataFrame(rows, columns=["year", "person_id", "company_id",
                                       "tie", *TERMS, *INTERACTIONS])


def fit_mple(design: pd.DataFrame, terms: list[str] | None = None,
             n_boot: int = 200, seed: int = 20260912) -> pd.DataFrame:
    """Pseudolikelihood fit with a node bootstrap, as `btergm` does by default.

    The bootstrap resamples **directors** with replacement and refits, which
    respects the fact that a director's dyads are not independent of each
    other. Resampling dyads would treat them as if they were and would give
    intervals that are far too narrow.
    """
    import warnings

    import statsmodels.api as sm

    terms = list(terms or TERMS)
    y = design.tie.to_numpy()
    X = design[terms].to_numpy(dtype=float)

    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        fit = sm.Logit(y, X).fit(disp=0, maxiter=200)

    rng = np.random.default_rng(seed)
    people = design.person_id.to_numpy()
    unique = np.unique(people)
    index: dict[str, np.ndarray] = {p: np.where(people == p)[0] for p in unique}
    draws = np.full((n_boot, len(terms)), np.nan)
    for b in range(n_boot):
        picked = rng.choice(unique, size=unique.size, replace=True)
        rows = np.concatenate([index[p] for p in picked])
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            try:
                draws[b] = sm.Logit(y[rows], X[rows]).fit(disp=0,
                                                          maxiter=200).params
            except Exception:            # a draw with no variation on a term
                continue

    ok = ~np.isnan(draws).any(axis=1)
    return pd.DataFrame({
        "term": terms,
        "estimate": fit.params,
        "boot_se": np.nanstd(draws[ok], axis=0, ddof=1),
        "lo": np.nanpercentile(draws[ok], 2.5, axis=0),
        "hi": np.nanpercentile(draws[ok], 97.5, axis=0),
        "n_dyads": len(design),
        "n_ties": int(y.sum()),
        "n_boot": int(ok.sum()),
    })


def formation_vs_retention(design: pd.DataFrame, n_boot: int = 100,
                           seed: int = 20260913) -> pd.DataFrame:
    """Split the model into the two processes a cross-section cannot separate.

    A TERGM's whole advantage over a cross-sectional ERGM is that it knows
    which ties are new. Fitting the same terms separately to the dyads that
    were **empty** at t-1 and to those that were **filled** asks two different
    questions: what draws a director onto a board he was not on, and what
    keeps him on one he was.

    `memory` is dropped from both — it is constant within each subset and is
    what defines them.
    """
    terms = [t for t in TERMS if t != "memory"]
    out = []
    for label, subset in (("formation", design[design.memory == 0]),
                          ("retention", design[design.memory == 1])):
        fitted = fit_mple(subset, terms=terms, n_boot=n_boot, seed=seed)
        fitted.insert(0, "process", label)
        out.append(fitted)
    return pd.concat(out, ignore_index=True)

File: scripts/tergm.R
#!/usr/bin/env Rscript
# Temporal ERGM for the Egyptian corporate elite network, 1932-1950.
#
# Fitted to the TWO-MODE network. Not to a projection: a one-mode projection
# of an affiliation network makes a clique out of every board, so its
# clustering is arithmetic rather than behaviour. Interlocking appears here as
# b2star(2) on the two-mode graph, which is the same quantity without the
# manufactured dependence.
#
# Composition change is handled by btergm's own adjustment, which conforms each
# time step to the nodes it shares with the previous one. It arrives at exactly
# the pairwise at-risk sets the Python fit in `politi.tergm` constructs by hand
# — 44x109, 152x228, 307x296, 459x427 — so the two are fitted to the same
# dyads and their coefficients are directly comparable.
#
# The union-plus-NA alternative was tried and abandoned: masking ~4.4 million
# non-existent dyads per wave stores them as missing edges and exhausts memory.
#
# Every term emits exactly one statistic. Factor terms are avoided because a
# level absent from one adjusted time step makes them emit different numbers
# of statistics per step, and btergm then fails in rbind. Binary covariates do
# not have that problem.
#
# Inputs are written by `politi.tergm.export_for_r`:
#   data/processed/tergm/tergm_edges.csv    year, person_id, company_id
#   data/processed/tergm/tergm_persons.csv  year, person_id, rank, origin, political
#   data/processed/tergm/tergm_firms.csv    year, company_id, sector
#
# Usage:  Rscript scripts/tergm.R [--from 1938] [--boot 500]
#         Rscript scripts/tergm.R --gof yes [--nsim 100]

suppressPackageStartupMessages({
  library(btergm)
  library(network)
})

args <- commandArgs(trailingOnly = TRUE)
opt <- function(flag, default) {
  i <- match(flag, args)
  if (is.na(i) || i == length(args)) default else args[i + 1]
}
first_wave <- as.integer(opt("--from", "1932"))
n_boot <- as.integer(opt("--boot", "500"))

root <- normalizePath(file.path(dirname(sub("--file=", "", grep("--file=",
  commandArgs(FALSE), value = TRUE)[1])), ".."), mustWork = FALSE)
if (!dir.exists(file.path(root, "data"))) root <- "."
indir <- file.path(root, "data", "processed", "tergm")

edges <- read.csv(file.path(indir, "tergm_edges.csv"), stringsAsFactors = FALSE)
persons <- read.csv(file.path(indir, "tergm_persons.csv"), stringsAsFactors = FALSE)
firms <- read.csv(file.path(indir, "tergm_firms.csv"), stringsAsFactors = FALSE)

waves <- sort(unique(edges$year))
waves <- waves[waves >= first_wave]
cat(sprintf("waves: %s   bootstrap replications: %d\n",
            paste(waves, collapse = ", "), n_boot))

# --- node attributes ---------------------------------------------------------
# Attributes are held on the union of nodes and indexed into each wave, so a
# director carries the same origin in every volume he appears in.
all_p <- sort(unique(persons$person_id[persons$year %in% waves]))
all_f <- sort(unique(firms$company_id[firms$year %in% waves]))
np <- length(all_p); nf <- length(all_f)
cat(sprintf("union node set: %d directors x %d firms\n", np, nf))

# Origin is a property of the man. Office is TRUE if recorded in any wave in
# range, which is the floor the annuaire supports.
attr_of <- function(ids, frame, key, column, fallback) {
  sub <- frame[frame$year %in% waves, ]
  sub <- sub[!duplicated(sub[[key]]), ]
  out <- sub[[column]][match(ids, sub[[key]])]
  out[is.na(out)] <- fallback
  out
}
p_origin <- attr_of(all_p, persons, "person_id", "origin", "unknown")
f_sector <- attr_of(all_f, firms, "company_id", "sector", "other")
# Office is WAVE-SPECIFIC, not a property of the man: a director recorded in
# office in 1947 was not necessarily in office in 1938. Collapsing it to
# "ever held office" both widens the group and would no longer match the
# Python fit the R run exists to check.
persons$office <- as.integer(persons$political %in% c("True", "TRUE", "true"))
#: Origin levels the homophily term counts. "unknown" is excluded: directors
#: whose origin could not be imputed are not a community and must not be
#: allowed to match each other.
origin_levels <- c("arab_egyptian", "european", "local_minority")
in_range <- persons$year %in% waves
cat(sprintf("office holders by wave: %s\n",
            paste(sprintf("%d:%d", sort(unique(persons$year[in_range])),
                          tapply(persons$office[in_range],
                                 persons$year[in_range], sum)),
                  collapse = " ")))

# --- one bipartite network per wave -----------------------------------------
# Built on the nodes present in that wave. btergm conforms the series itself.
make_net <- function(y) {
  p <- sort(unique(persons$person_id[persons$year == y]))
  f <- sort(unique(firms$company_id[firms$year == y]))
  m <- matrix(0L, nrow = length(p), ncol = length(f), dimnames = list(p, f))
  e <- edges[edges$year == y, ]
  m[cbind(match(e$person_id, p), match(e$company_id, f))] <- 1L
  n <- network(m, bipartite = length(p), directed = FALSE,
               matrix.type = "bipartite")
  pi <- match(p, all_p); fi <- match(f, all_f)
  # Origin stays categorical: homophily needs it. Everything else is binary,
  # so each term contributes one statistic in every time step.
  set.vertex.attribute(n, "origin",
                       c(p_origin[pi], rep("firm", length(f))))
  wave_office <- persons[persons$year == y, ]
  set.vertex.attribute(n, "office",
                       c(wave_office$office[match(p, wave_office$person_id)],
                         rep(0L, length(f))))
  for (sec in c("finance", "land_property", "agriculture")) {
    set.vertex.attribute(n, sec, c(rep(0L, length(p)),
                                   as.integer(f_sector[fi] == sec)))
  }
  n
}

nets <- lapply(waves, make_net)
names(nets) <- as.character(waves)
for (i in seq_along(nets)) {
  cat(sprintf("  %s: %d directors x %d firms, %d ties\n", names(nets)[i],
              nets[[i]] %n% "bipartite",
              network.size(nets[[i]]) - (nets[[i]] %n% "bipartite"),
              network.edgecount(nets[[i]])))
}

# --- the model ---------------------------------------------------------------
# Only attributes read off the page are covariates. Seat counts, financier
# status and centrality are functions of the ties being modelled, and putting
# any of them on the right-hand side would regress the network on itself.
set.seed(20260912)
model <- btergm(
  nets ~ edges +
    memory(type = "autoregression") +          # a seat held again
    b1star(2) +                                # a director holding two seats
    b2star(2) +                                # two directors sharing a firm
    b1cov("office") +                          # office holders
    b1nodematch("origin", levels = I(origin_levels)) +
    b2cov("finance") + b2cov("land_property") + b2cov("agriculture"),
  R = n_boot, parallel = "no", verbose = TRUE
)

cat("\n==================== TERGM ====================\n")
print(summary(model))

# btergm's confint returns four columns — Estimate, Boot mean, 2.5%, 97.5% —
# so the interval must be selected by name. Taking the first two columns
# silently writes the estimate and the bootstrap mean as if they were bounds.
ci <- confint(model)
pct <- grep("%", colnames(ci))
stopifnot(length(pct) == 2)
out <- data.frame(term = rownames(ci), estimate = coef(model),
                  boot_mean = ci[, "Boot mean"],
                  lo = ci[, pct[1]], hi = ci[, pct[2]], row.names = NULL)
outdir <- file.path(root, "data", "processed")
tag <- if (first_wave > 1932) sprintf("_from%d", first_wave) else ""
write.csv(out, file.path(outdir, sprintf("tergm_coefficients%s.csv", tag)),
          row.names = FALSE)
cat(sprintf("\nwrote %s\n",
            file.path(outdir, sprintf("tergm_coefficients%s.csv", tag))))

# --- goodness of fit -----------------------------------------------------------
# Reported because a TERGM that misses the degree distribution of an affiliation
# network is describing something other than this network.
#
# The statistics are chosen for a BIPARTITE graph:
#   b1deg, b2deg  the two degree distributions -- how many seats a director
#                 holds and how many directors a firm has. These are what the
#                 b1star(2) and b2star(2) terms exist to reproduce, so they are
#                 the direct test of the terms this model leans on.
#   dsp           dyadwise shared partners. In a two-mode network this is the
#                 closure statistic: two directors sharing boards, two firms
#                 sharing directors. The model has NO closure term, and
#                 `docs/EMBEDDEDNESS.md` reports that closure is the largest
#                 predictor of tie formation in this data, so this is where the
#                 specification is most likely to fail -- which is the reason
#                 to look rather than a reason not to.
#   geodesic      the distance distribution, i.e. whether the simulated network
#                 has this one's reach.
#   rocpr         tie prediction, ROC and precision-recall. At a density under
#                 0.5% the PR curve is the informative one; ROC will look good
#                 whatever the model does.
#
# `esp` (edgewise shared partners) is deliberately NOT included. It counts
# partners shared across an existing edge, which requires a triangle, and a
# bipartite graph has none: it would report a column of zeros matched by a
# column of zeros and read as a perfect fit.
if (identical(opt("--gof", "yes"), "yes")) {
  cat("\n---- goodness of fit ----\n")
  nsim <- as.integer(opt("--nsim", "100"))
  set.seed(20260915)
  g <- gof(model, statistics = c(b1deg, b2deg, dsp, geodesic, rocpr),
           nsim = nsim)
  print(g)

  gofdir <- file.path(outdir, sprintf("tergm_gof%s", tag))
  dir.create(gofdir, recursive = TRUE, showWarnings = FALSE)
  # The distribution statistics carry a $stats data frame: observed and
  # simulated mean/median/min/max per level, and a p-value. `rocpr` does not --
  # it holds AUCs and curves instead -- so it is written separately rather than
  # skipped, which an earlier version of this block did silently.
  for (stat in names(g)) {
    entry <- g[[stat]]
    slug <- gsub("^_|_$", "", gsub("[^a-z0-9]+", "_", tolower(stat)))
    if (!is.null(entry$stats)) {
      write.csv(data.frame(statistic = stat, level = rownames(entry$stats),
                           entry$stats, check.names = FALSE),
                file.path(gofdir, sprintf("%s.csv", slug)), row.names = FALSE)
    } else if (!is.null(entry$auc.roc)) {
      # auc.*.rgraph is the same AUC for a random graph of the same density:
      # the baseline the model has to beat. At this density the PR area is the
      # informative one; ROC is high for almost any model.
      write.csv(data.frame(
        statistic = stat,
        measure = c("auc.roc", "auc.roc.rgraph", "auc.pr", "auc.pr.rgraph"),
        value = c(mean(entry$auc.roc), mean(entry$auc.roc.rgraph),
                  mean(entry$auc.pr), mean(entry$auc.pr.rgraph))),
        file.path(gofdir, sprintf("%s.csv", slug)), row.names = FALSE)
    }
  }
  pdf(file.path(gofdir, "gof.pdf"), width = 9, height = 6)
  plot(g)
  dev.off()
  cat(sprintf("\nwrote %s\n", gofdir))
}
